Обменный курс

licosty opened the pull request:
[Задача JIRA](https://faang-school.atlassian.net/browse/BJS2-83795)

licosty committed:
currency: fetch rates, cache

licosty committed:
currency: tests

zaqqaz226 commented on src/main/java/faang/school/paymentservice/client/currency/ExchangeRatesApi.java:
```diff
@@ -0,0 +1,97 @@
+        return exchangeRatesWebClient.get()
+                .uri(this::buildUri)
+                .retrieve()
+                .onStatus(HttpStatusCode::is4xxClientError, this::processHttpErrorCodeResponse)
```
Там вроде есть что то типа isError или что то такое. И на все не ОК ответы будет обработка.

licosty replied:
Да, переделала, теперь лучше)

zaqqaz226 commented on src/main/java/faang/school/paymentservice/client/currency/ExchangeRatesApi.java:
```diff
@@ -0,0 +1,97 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ExchangeRatesApi implements CurrencyClient {
```
<img width="1500" height="844" alt="image" src="https://github.com/user-attachments/assets/3c145b6c-6997-4991-871a-3050b0fd02b5" />
Вау!)

zaqqaz226 commented on src/main/java/faang/school/paymentservice/client/currency/ExchangeRatesApi.java:
```diff
@@ -0,0 +1,97 @@
+        log.info("Processing response from API: {}", response);
+        if (!response.success()) {
+            if (response.error() == null) {
+                return Mono.error(new InvalidApiResponseException("Field 'success' is false, but error is null"));
```
Проще отдавать в ошибку какой нить Enum. В котором будет нужный код, ошибка и http код, который надо будет отдать на хендлере.

licosty replied:
Здесь не совсем поняла.
Имеешь ввиду что-то вроде такого?
```
enum ApiErrorType {
    RESOURCE_NOT_EXIST(404, "Some error text", HttpStatus.NOT_FOUND);
}
```
и потом 
`
return Mono.error(new InvalidApiResponseException(ApiErrorType.getByCode(response.error().code()));
`

zaqqaz226 replied:
InvalidApiResponseException это ж твое собственное исключение. Значит в его конструктор можно засунуть что захочешь)) Скажи, чтобы оно ожидало твой Enum. И при выбросе просто отдавай туда нужный Enum. А на GlobalExcHandler`е будешь ловить это исключение, забирать из него Enum и формировать нужный тебе ответ с нужным кодом.

licosty replied:
Не могу понять чем хуже отдавать ошибку просто текстом

zaqqaz226 replied:
Не хуже, но ты ж спрашивала как сделать код более расширяемым) У тебя в ситуации с Enum проще отдавать ошибки, проще расширять, проще использовать. ВСе что надо будет, это дописать значение в enum и будет отдана нужная ошибка с нужным кодом. И если надо будет что-то поменять в ошибке, скажем хттп код, то тебе не надо лазить по всем и искать, ты просто поменяешь в одном месте

licosty replied:
Сделала. Вроде такого?)

zaqqaz226 commented on src/main/java/faang/school/paymentservice/client/currency/ExchangeRatesApi.java:
```diff
@@ -0,0 +1,97 @@
+
+    private URI buildUri(UriBuilder uriBuilder) {
+        String currencies = Currency.getCurrenciesAsString();
+        log.info("Currency list: {}", currencies);
```
Сообщение лога я бы доработал. А то если представить, что ты его читаешь в простыне остальных логов, не понятно зачем он

licosty replied:
Вроде сделала получше)

zaqqaz226 commented on src/main/java/faang/school/paymentservice/client/currency/ExchangeRatesApi.java:
```diff
@@ -0,0 +1,97 @@
+                .build();
+    }
+
+    private Retry prepareBehaviorRetry() {
```
<img width="1200" height="569" alt="image" src="https://github.com/user-attachments/assets/2228f77b-504d-4a5d-ae7a-305bcee118cc" />

zaqqaz226 replied:
Но вопрос тебе на засыпку) А не стоит ли это вынести в конфигурацию бина?) (в вопросе нету правильного ответа))

licosty replied:
Вот смотрела я на этот метод и что-то было не к душе)
Про бин не догадалась, зато теперь лучше)

zaqqaz226 commented on src/main/java/faang/school/paymentservice/config/WebClientConfig.java:
```diff
@@ -0,0 +1,23 @@
+    private final ExchangeRatesProperty exchangeRatesProperty;
+
+    @Bean
+    public WebClient exchangeRatesWebClient() {
```
```
  @Bean
  public WebClient webClient() {
    HttpClient client = HttpClient
        .create()
        .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, requestProperties.connectTimeoutMs())
        .responseTimeout(Duration.ofMillis(requestProperties.responseTimeoutMs()));
    return WebClient
        .builder()
        .clientConnector(new ReactorClientHttpConnector(client))
        .build();
  }
```

Лушче как нить так. Мы тут сразу делаем опции по отваливанию при реквесте и опция по отваливанию при ожидании ответа. Таймауты берем из пропертей конечно же)

licosty replied:
Добавила

licosty committed:
currency:
- оставлен один onStatus для не OK статусов
- retry вынесен как бин
- таймауты в WebClient
- скорректированы тесты

licosty committed:
currency:
- enam ErrorType для фикс.ошибок
